BxlAsLib.txt

```diff
@@ -0,0 +1,377 @@
+  - Creating, serializing, and deserializing pips must be performant.
+  - Stitching multiple fragments must be performant.
+
+- Graph fragment interactions:
+  - Fragment can consume artifacts produced by other frontends.
+  - Fragment can produce artifacts that can be consumed by other frontends.
+
+- API for constructing graph fragments that makes sense for customer
+  - Adding basic pips, e.g., process, copy, write, seal directory.
+  - Extensible for adding BuildXL's friends:
+    - Adding QTest pips.
+    - Adding drop and symbol pips.    
+  - Desirable:
+    - Higher-level API to ease migration from existing DScript SDK.
+      - Creating a unique output directory.
+      - Specifying inputs/outputs of pips like in DScript.
+    - API becomes yet another frontend.
+
+- Reuse existing DScript (or other frontends) SDK: Drop, Symbol, QTest, etc.
+
+- Control over BuildXL's friends behavior, e.g., adding a new default to QTest pips.
+
+- Track what files and environment variables used for graph fragment constructions.
+  - This is needed in case the construction is not done in a sandboxed environment.
+
+- BXLLib is applicable to many customers, not only to Office, e.g., WDG can benefit from BXLLib.
+
+Spike
+-----
+- Use BuildXL DScript frontend to evaluate specs and generate specific fragments.
+  - Description:
+    - This approach targets only performance.
+    - In metabuild, besides generating DScript spec, the pip also evaluates the spec (plus the SDK) and generates a fragment.
+    - Instead of using a full-blown BuildXL, we refactor BuildXL frontend to create so-called BXLScriptAnalyzer
+      that is used for evaluation and for generating fragment.
+    - The generated fragments are then fed as inputs to product build.
+    - Performance can be gained because:
+      - The metabuild pip generates a smaller spec that can be evaluated fast (to be proven).
+      - The metabuild pip can be distributed. 
+
+  - Pros:
+    - It's a shortcut and requires moderate efforts.
+    - We can (hopefully) deliver performance without changing DScript SDK, without changing static NMake, without
+      waiting for definite API.
+    - Existing DScript SDK can still be used.
+    - The shipment of BuildXL engine and its DScript SDK doesn't change.
+
+  - Cons:
+    - This may be applicable to Office where their project slices are kind of isolated from one another.
+    - Stitching fragments is non-trivial due to service and IPC pips:
+      - [In progress] Need to unify service and IPC pips:
+        - Instead of having one drop service pip for the whole graph, we now have N drop service pips, where
+          N is the number of generated fragments.
+        - IPC pips are tied to specific service pips through monikers. We need to mess around with IPC pips
+          command-line arguments to unify service and IPC pips.
+
+API
+---
+- Basic API: must be small and not prone to binary incompatibility.

```

Comment:
How can an API be prone or not prone to binary incompatibility?  And why in particular the API proposed here is not prone to binary incompatibility?